File: implementations/java/common/src/test/java/com/theshuai/common/peeregress/PeerEgressVectorTests.java
package com.theshuai.common.peeregress;

import com.fasterxml.jackson.databind.JsonNode;
import com.theshuai.common.util.JsonUtil;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Binds this implementation to the shared vectors under {@code protocol/test-vectors}.
 *
 * <p>The assertions check the returned code, not just allow versus deny: a runtime that denies for
 * the wrong reason would still leave operators unable to tell a revoked ACL from a closed port.
 */
class PeerEgressVectorTests {
    @Test
    void authorizationMatchesSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-authz-v1.json");
        PeerEgressPolicy basePolicy = policyOf(vector.path("policy"));
        PeerEgressAuthorization.Context context = contextOf(vector);
        int checked = 0;
        for (JsonNode node : vector.path("cases")) {
            PeerEgressRequest request = requestOf(node.path("request"));
            PeerEgressAuthorization.Decision decision =
                    PeerEgressAuthorization.evaluate(request, basePolicy, true, context);
            String name = node.path("name").asText();
            assertEquals(node.path("expect").path("code").asText(), decision.code(), name);
            assertEquals(node.path("expect").path("allowed").asBoolean(), decision.allowed(), name);
            checked++;
        }
        assertTrue(checked > 0, "authorization vector carried no cases");
    }

    /**
     * Boundary sweep shared by every runtime.
     *
     * <p>The hand-written cases document intent; this exists to catch a runtime that agrees on the
     * documented examples but diverges one address or one port away from a boundary. Every runtime
     * runs the same sweep against the same policy, so a disagreement here is a disagreement about
     * the rules rather than about the fixture.
     */
    @Test
    void crossLanguageBoundarySweepMatchesSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-authz-v1.json");
        PeerEgressPolicy basePolicy = policyOf(vector.path("policy"));
        PeerEgressAuthorization.Context context = contextOf(vector);
        int checked = 0;
        for (JsonNode node : vector.path("crossLanguageCases")) {
            PeerEgressRequest request = requestOf(node.path("request"));
            PeerEgressAuthorization.Decision decision =
                    PeerEgressAuthorization.evaluate(request, basePolicy, true, context);
            String name = node.path("name").asText();
            assertEquals(node.path("expect").path("code").asText(), decision.code(), name);
            assertEquals(node.path("expect").path("allowed").asBoolean(), decision.allowed(), name);
            checked++;
        }
        assertTrue(checked >= 100, "cross-language sweep carried only " + checked + " cases");
    }

    @Test
    void policyVariantsMatchSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-authz-v1.json");
        PeerEgressAuthorization.Context context = contextOf(vector);
        int checked = 0;
        for (JsonNode node : vector.path("policyVariantCases")) {
            PeerEgressPolicy policy = policyOf(vector.path("policy"));
            JsonNode override = node.path("policyOverride");
            if (override.isObject()) {
                if (override.has("enabled")) {
                    policy.setEnabled(override.path("enabled").asBoolean());
                }
                if (override.has("scope")) {
                    policy.setScope(override.path("scope").asText());
                }
                if (override.has("destinationRules")) {
                    policy.setDestinationRules(destinationRulesOf(override.path("destinationRules")));
                }
            }
            boolean peerAclAllows = !node.has("peerAclAllows") || node.path("peerAclAllows").asBoolean();
            PeerEgressRequest request = requestOf(node.path("request"));
            PeerEgressAuthorization.Decision decision =
                    PeerEgressAuthorization.evaluate(request, policy, peerAclAllows, context);
            String name = node.path("name").asText();
            assertEquals(node.path("expect").path("code").asText(), decision.code(), name);
            checked++;
        }
        assertTrue(checked > 0, "authorization vector carried no policy variants");
    }

    /**
     * The vector's policy deliberately contains a {@code 0.0.0.0/0} rule. Every address the vector
     * declares as forced-deny must still be refused through it.
     */
    @Test
    void forcedDenyDestinationsSurviveABroadAllowRule() throws IOException {
        JsonNode vector = readVector("peer-egress-authz-v1.json");
        PeerEgressPolicy policy = policyOf(vector.path("policy"));
        PeerEgressAuthorization.Context context = contextOf(vector);
        assertTrue(policy.getDestinationRules().stream().anyMatch(rule -> "0.0.0.0/0".equals(rule.getCidr())),
                "vector policy must keep a broad rule, otherwise this test proves nothing");

        List<String> denied = new ArrayList<>();
        for (JsonNode node : vector.path("forcedDenyCidrs")) {
            denied.add(node.asText());
        }
        for (JsonNode node : vector.path("cloudMetadataCidrs")) {
            denied.add(node.asText());
        }
        assertFalse(denied.isEmpty(), "vector declared no forced-deny ranges");

        for (String text : denied) {
            Ipv4Cidr cidr = Ipv4Cidr.parse(text);
            assertNotNull(cidr, text);
            PeerEgressRequest request = new PeerEgressRequest();
            request.setConsumerClientId(1);
            request.setDestinationIp(Ipv4Cidr.format(cidr.network()));
            request.setDestinationPort(443);
            request.setProtocol("tcp");
            PeerEgressAuthorization.Decision decision =
                    PeerEgressAuthorization.evaluate(request, policy, true, context);
            assertEquals(PeerEgressCodes.FORBIDDEN_DESTINATION, decision.code(), text);
        }
    }

    @Test
    void ruleMatchingMatchesSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-rules-v1.json");
        List<PeerEgressRule> rules = new ArrayList<>();
        for (JsonNode node : vector.path("rules")) {
            rules.add(JsonUtil.stringToObject(node.toString(), PeerEgressRule.class));
        }
        assertFalse(rules.isEmpty(), "rules vector carried no rules");

        for (JsonNode node : vector.path("cases")) {
            String name = node.path("name").asText();
            PeerEgressRules.Match match = PeerEgressRules.match(rules, node.path("destination").asText());
            JsonNode expect = node.path("expect");
            assertEquals(expect.path("action").asText(), match.action(), name);
            if (expect.path("matchedRuleIndex").isNull()) {
                assertNull(match.matchedRuleIndex(), name);
                assertFalse(match.matched(), name);
            } else {
                assertEquals(expect.path("matchedRuleIndex").asInt(), match.matchedRuleIndex(), name);
            }
            if (expect.has("egressClientId")) {
                assertEquals(expect.path("egressClientId").asLong(), match.egressClientId(), name);
            }
        }
    }

    @Test
    void ruleValidationMatchesSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-rules-v1.json");
        String meshCidr = vector.path("meshCidr").asText(PeerEgressRules.DEFAULT_MESH_CIDR);

        for (JsonNode node : vector.path("rules")) {
            PeerEgressRule rule = JsonUtil.stringToObject(node.toString(), PeerEgressRule.class);
            assertNull(PeerEgressRules.validate(rule, meshCidr),
                    "rule " + node.path("match").asText() + " must pass validation");
        }

        int checked = 0;
        for (JsonNode node : vector.path("configValidation")) {
            PeerEgressRule rule = JsonUtil.stringToObject(node.path("rule").toString(), PeerEgressRule.class);
            assertEquals(node.path("code").asText(), PeerEgressRules.validate(rule, meshCidr),
                    node.path("name").asText());
            checked++;
        }
        assertTrue(checked > 0, "rules vector carried no configValidation cases");
    }

    private static PeerEgressAuthorization.Context contextOf(JsonNode vector) {
        String meshCidr = PeerEgressRules.DEFAULT_MESH_CIDR;
        for (JsonNode node : vector.path("forcedDenyCidrs")) {
            Ipv4Cidr cidr = Ipv4Cidr.parse(node.asText());
            if (cidr != null && cidr.prefixLength() == 11) {
                meshCidr = node.asText();
            }
        }
        return new PeerEgressAuthorization.Context(meshCidr, List.of());
    }

    private static PeerEgressPolicy policyOf(JsonNode node) {
        return JsonUtil.stringToObject(node.toString(), PeerEgressPolicy.class);
    }

    private static PeerEgressRequest requestOf(JsonNode node) {
        return JsonUtil.stringToObject(node.toString(), PeerEgressRequest.class);
    }

    private static List<PeerEgressPolicy.PeerEgressDestinationRule> destinationRulesOf(JsonNode node) {
        List<PeerEgressPolicy.PeerEgressDestinationRule> rules = new ArrayList<>();
        for (JsonNode entry : node) {
            rules.add(JsonUtil.stringToObject(entry.toString(),
                    PeerEgressPolicy.PeerEgressDestinationRule.class));
        }
        return rules;
    }

    private static JsonNode readVector(String fileName) throws IOException {
        Path current = Path.of("").toAbsolutePath();
        for (int depth = 0; current != null && depth < 8; depth++, current = current.getParent()) {
            Path candidate = current.resolve("protocol/test-vectors/" + fileName);
            if (Files.isRegularFile(candidate)) {
                return JsonUtil.readString(Files.readString(candidate));
            }
        }
        throw new IllegalStateException("cannot locate " + fileName);
    }
}

File: implementations/csharp/protocol/tests/Specus.Protocol.Tests/PeerEgressVectorTests.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Specus.Protocol.PeerEgress;

namespace Specus.Protocol.Tests;

/// <summary>
/// Binds this implementation to the shared vectors under protocol/test-vectors.
/// </summary>
/// <remarks>
/// The assertions check the returned code, not just allow versus deny: a runtime that denies for
/// the wrong reason still leaves an operator unable to tell a revoked ACL from a closed port.
/// </remarks>
public sealed class PeerEgressVectorTests
{
    [Fact]
    public void AuthorizationMatchesSharedVector()
    {
        var vector = ReadVector<AuthzVector>("peer-egress-authz-v1.json");
        Assert.NotEmpty(vector.Cases);
        var context = ContextFor(vector);
        foreach (var testCase in vector.Cases)
        {
            var decision = PeerEgressAuthorization.Authorize(testCase.Request, vector.Policy, true, context);
            Assert.Equal(testCase.Expect.Code, decision.Code);
            Assert.Equal(testCase.Expect.Allowed, decision.Allowed);
        }
    }

    [Fact]
    public void PolicyVariantsMatchSharedVector()
    {
        var vector = ReadVector<AuthzVector>("peer-egress-authz-v1.json");
        Assert.NotEmpty(vector.PolicyVariantCases);
        var context = ContextFor(vector);
        foreach (var testCase in vector.PolicyVariantCases)
        {
            var policy = vector.Policy;
            var overrides = testCase.PolicyOverride;
            if (overrides is not null)
            {
                policy = policy with
                {
                    Enabled = overrides.Enabled ?? policy.Enabled,
                    Scope = overrides.Scope ?? policy.Scope,
                    DestinationRules = overrides.DestinationRules ?? policy.DestinationRules,
                };
            }
            var decision = PeerEgressAuthorization.Authorize(
                testCase.Request, policy, testCase.PeerAclAllows ?? true, context);
            Assert.Equal(testCase.Expect.Code, decision.Code);
        }
    }

    /// <summary>
    /// The vector policy deliberately contains a 0.0.0.0/0 rule. Every address the vector declares
    /// as forced-deny must still be refused through it.
    /// </summary>
    [Fact]
    public void ForcedDenyDestinationsSurviveABroadAllowRule()
    {
        var vector = ReadVector<AuthzVector>("peer-egress-authz-v1.json");
        Assert.Contains(vector.Policy.DestinationRules, rule => rule.Cidr == "0.0.0.0/0");

        var context = ContextFor(vector);
        var denied = vector.ForcedDenyCidrs.Concat(vector.CloudMetadataCidrs).ToArray();
        Assert.NotEmpty(denied);
        foreach (var text in denied)
        {
            Assert.True(Ipv4Cidr.TryParse(text, out var cidr), text);
            var decision = PeerEgressAuthorization.Authorize(
                new PeerEgressRequest
                {
                    ConsumerClientId = 1,
                    DestinationIp = Ipv4Cidr.FormatAddress(cidr.Network),
                    DestinationPort = 443,
                    Protocol = "tcp",
                },
                vector.Policy, true, context);
            Assert.Equal(PeerEgressCodes.ForbiddenDestination, decision.Code);
        }
    }

    [Fact]
    public void RuleMatchingMatchesSharedVector()
    {
        var vector = ReadVector<RulesVector>("peer-egress-rules-v1.json");
        Assert.NotEmpty(vector.Rules);
        foreach (var testCase in vector.Cases)
        {
            var match = PeerEgressRules.Match(vector.Rules, testCase.Destination);
            Assert.Equal(testCase.Expect.Action, match.Action);
            if (testCase.Expect.MatchedRuleIndex is null)
            {
                Assert.False(match.Matched, testCase.Name);
                continue;
            }
            Assert.Equal(testCase.Expect.MatchedRuleIndex.Value, match.MatchedRuleIndex);
            if (testCase.Expect.EgressClientId is not null)
            {
                Assert.Equal(testCase.Expect.EgressClientId, match.EgressClientId);
            }
        }
    }

    [Fact]
    public void RuleValidationMatchesSharedVector()
    {
        var vector = ReadVector<RulesVector>("peer-egress-rules-v1.json");
        var mesh = string.IsNullOrWhiteSpace(vector.MeshCidr) ? PeerEgressRules.DefaultMeshCidr : vector.MeshCidr;
        foreach (var rule in vector.Rules)
        {
            Assert.Null(PeerEgressRules.Validate(rule, mesh));
        }
        Assert.NotEmpty(vector.ConfigValidation);
        foreach (var testCase in vector.ConfigValidation)
        {
            Assert.Equal(testCase.Code, PeerEgressRules.Validate(testCase.Rule, mesh));
        }
    }

    /// <summary>
    /// Boundary sweep shared by every runtime.
    /// </summary>
    /// <remarks>
    /// The hand-written cases document intent; this exists to catch a runtime that agrees on the
    /// documented examples but diverges one address or one port away from a boundary. Every runtime
    /// runs the same sweep against the same policy, so a disagreement here is a disagreement about
    /// the rules rather than about the fixture.
    /// </remarks>
    [Fact]
    public void CrossLanguageBoundarySweepMatchesSharedVector()
    {
        var vector = ReadVector<AuthzVector>("peer-egress-authz-v1.json");
        Assert.True(vector.CrossLanguageCases.Count >= 100,
            $"cross-language sweep carried only {vector.CrossLanguageCases.Count} cases");
        var context = ContextFor(vector);
        foreach (var testCase in vector.CrossLanguageCases)
        {
            var decision = PeerEgressAuthorization.Authorize(testCase.Request, vector.Policy, true, context);
            Assert.Equal(testCase.Expect.Code, decision.Code);
            Assert.Equal(testCase.Expect.Allowed, decision.Allowed);
        }
    }

    /// <summary>
    /// The mesh network reaches the implementation through the deployment context rather than the
    /// static list, so it is lifted back out of the vector forced-deny set by its prefix length.
    /// </summary>
    private static PeerEgressContext ContextFor(AuthzVector vector)
    {
        var mesh = PeerEgressRules.DefaultMeshCidr;
        foreach (var text in vector.ForcedDenyCidrs)
        {
            if (Ipv4Cidr.TryParse(text, out var cidr) && cidr.PrefixLength == 11)
            {
                mesh = text;
            }
        }
        return new PeerEgressContext { MeshCidr = mesh };
    }

    private static T ReadVector<T>(string name)
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        for (var depth = 0; directory is not null && depth < 12; depth++, directory = directory.Parent)
        {
            var candidate = Path.Combine(directory.FullName, "protocol", "test-vectors", name);
            if (!File.Exists(candidate))
            {
                continue;
            }
            return JsonSerializer.Deserialize<T>(File.ReadAllText(candidate))
                ?? throw new InvalidDataException($"{name} decoded to null");
        }
        throw new FileNotFoundException($"cannot locate {name}");
    }

    private sealed record Expectation
    {
        [JsonPropertyName("allowed")]
        public bool Allowed { get; init; }

        [JsonPropertyName("code")]
        public string Code { get; init; } = string.Empty;
    }

    private sealed record AuthzVector
    {
        [JsonPropertyName("forcedDenyCidrs")]
        public IReadOnlyList<string> ForcedDenyCidrs { get; init; } = [];

        [JsonPropertyName("cloudMetadataCidrs")]
        public IReadOnlyList<string> CloudMetadataCidrs { get; init; } = [];

        [JsonPropertyName("lanCidrs")]
        public IReadOnlyList<string> LanCidrs { get; init; } = [];

        [JsonPropertyName("policy")]
        public PeerEgressPolicy Policy { get; init; } = new();

        [JsonPropertyName("cases")]
        public IReadOnlyList<AuthzCase> Cases { get; init; } = [];

        [JsonPropertyName("crossLanguageCases")]
        public IReadOnlyList<AuthzCase> CrossLanguageCases { get; init; } = [];

        [JsonPropertyName("policyVariantCases")]
        public IReadOnlyList<PolicyVariantCase> PolicyVariantCases { get; init; } = [];
    }

    private sealed record AuthzCase
    {
        [JsonPropertyName("name")]
        public string Name { get; init; } = string.Empty;

        [JsonPropertyName("request")]
        public PeerEgressRequest Request { get; init; } = new();

        [JsonPropertyName("expect")]
        public Expectation Expect { get; init; } = new();
    }

    private sealed record PolicyOverride
    {
        [JsonPropertyName("enabled")]
        public bool? Enabled { get; init; }

        [JsonPropertyName("scope")]
        public string? Scope { get; init; }

        [JsonPropertyName("destinationRules")]
        public IReadOnlyList<PeerEgressDestinationRule>? DestinationRules { get; init; }
    }

    private sealed record PolicyVariantCase
    {
        [JsonPropertyName("name")]
        public string Name { get; init; } = string.Empty;

        [JsonPropertyName("policyOverride")]
        public PolicyOverride? PolicyOverride { get; init; }

        [JsonPropertyName("peerAclAllows")]
        public bool? PeerAclAllows { get; init; }

        [JsonPropertyName("request")]
        public PeerEgressRequest Request { get; init; } = new();

        [JsonPropertyName("expect")]
        public Expectation Expect { get; init; } = new();
    }

    private sealed record RulesVector
    {
        [JsonPropertyName("meshCidr")]
        public string MeshCidr { get; init; } = string.Empty;

        [JsonPropertyName("rules")]
        public IReadOnlyList<PeerEgressRule> Rules { get; init; } = [];

        [JsonPropertyName("cases")]
        public IReadOnlyList<RuleCase> Cases { get; init; } = [];

        [JsonPropertyName("configValidation")]
        public IReadOnlyList<ValidationCase> ConfigValidation { get; init; } = [];
    }

    private sealed record RuleCase
    {
        [JsonPropertyName("name")]
        public string Name { get; init; } = string.Empty;

        [JsonPropertyName("destination")]
        public string Destination { get; init; } = string.Empty;

        [JsonPropertyName("expect")]
        public RuleExpectation Expect { get; init; } = new();
    }

    private sealed record RuleExpectation
    {
        [JsonPropertyName("action")]
        public string Action { get; init; } = string.Empty;

        [JsonPropertyName("matchedRuleIndex")]
        public int? MatchedRuleIndex { get; init; }

        [JsonPropertyName("egressClientId")]
        public long? EgressClientId { get; init; }
    }

    private sealed record ValidationCase
    {
        [JsonPropertyName("name")]
        public string Name { get; init; } = string.Empty;

        [JsonPropertyName("rule")]
        public PeerEgressRule Rule { get; init; } = new();

        [JsonPropertyName("code")]
        public string Code { get; init; } = string.Empty;
    }
}
